### PR TITLE
[Fix] Use a safe stop-string decode window

### LINKED ISSUE #36698: [Bug] DeepSeek-V4-Flash-0731 fails to stop at stop sequence when preceded by Chinese text
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

We are observing inconsistent stop-sequence behavior when serving deepseek-v4-flash-0731 with SGLang in PD disaggregated.

When the configured stop sequence (`<END>`) is preceded by English text, SGLang correctly detects the stop sequence and truncates the output.

However, when the same stop sequence is preceded by Chinese text, the stop sequence is not detected/truncated, and the model continues generating text after it.

### Reproduction

Case 1: English text before the stop sequence:
```
curl -sS '127.0.0.1:8080' \
  -H "Authorization: Bearer <key>" \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "deepseek-v4-flash-0731",
    "messages": [
      {
        "role": "user",
        "content": "please output banana<END>apple, without any other explnation."
      }
    ],
    "max_tokens": 64,
    "temperature": 0,
    "stop": ["<END>"]
  }'
```

Output:
```
banana
```
This is the expected behavior. <END> is detected and the response is truncated before the stop sequence.

Case 2: Chinese text before the same stop sequence
```
curl -sS '127.0.0.1:8080' \
  -H "Authorization: Bearer <key>" \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "deepseek-v4-flash-0731",
    "messages": [
      {
        "role": "user",
        "content": "请输出香蕉<END>苹果，不要附带其他解释."
      }
    ],
    "max_tokens": 64,
    "temperature": 0,
    "stop": ["<END>"]
  }'
```

output
```
香蕉<END>苹果
```

### Environment

H20
slgang 0.5.17

### PR DESCRIPTION
**Pull request by DylanB03**
## Motivation

Fixes #36698.

Stop-string matching decodes a bounded tail of generated tokens. The bound was derived only from encoding the stop string in isolation, but generated text can use a longer non-canonical tokenization. For DeepSeek-V4-Flash-0731, `<END>` canonically encodes to 3 tokens but can be generated as 5 tokens, causing the decode window to retain only `END>` and miss the stop.

## Modifications

- Size the stop-string window using the larger of the canonical token count and UTF-8 byte length.
- Consolidate the two existing length tests into one regression test covering both sides of that bound.
- Update the existing serialization assertion for the corrected normalized value.

## Accuracy Tests

This does not change model forward computation or numerical accuracy.

- `pytest test/registered/unit/sampling/test_sampling_params.py -q`: 70 passed, 13 subtests passed.
- `pytest test/registered/unit/managers/test_stop_str_speculative.py -q`: 9 passed.
- DeepSeek-V4-Flash-0731 tokenizer check: canonical `<END>` IDs `[30, 16860, 32]`; non-canonical IDs `[30, 39, 48, 38, 32]`; corrected window decodes `香蕉<END>` and contains the stop string.

## Speed Tests and Profiling

Not run; this is a correctness fix. Normalization adds one UTF-8 encoding per configured stop string, and the existing tail decoder reads up to the corrected bound only when it is larger than the canonical token count.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). (No documentation change is needed for this internal correctness fix.)
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). (Not applicable; rationale and proportional correctness checks are reported above.)
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #33114659412](https://github.com/sgl-project/sglang/actions/runs/33114659412)
Latest PR Test (Extra): :x: [Run #33114659213](https://github.com/sgl-project/sglang/actions/runs/33114659213)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33114659289](https://github.com/sgl-project/sglang/actions/runs/33114659289)

### CODE REVIEW

**Comment by DylanB03** [comment]
Real-model A/B validation (single NVIDIA B300, 2026-08-28 UTC)

Tested exact revisions:

- baseline: `ff5578eb4e1d68d842303e3ab9ce4c86038c1c26`
- patch: `18edeb540b436f7a4d2de5fab006f7a314f4895a`
- model: `deepseek-ai/DeepSeek-V4-Flash-0731`, HF snapshot `7872f01b1d1fe23eabc4c98b48bffcef5a386062`
- image: `lmsysorg/sglang:v0.5.18` (`sha256:9e148f5ac788e856a06166bd6347a831831eb9fcfab4d1770874823a7c29a1a1`)
- runtime: Python 3.12.3, PyTorch 2.13.0+cu130, CUDA 13.0.88, driver 580.173.02

Both legs used the same server arguments and imported `sampling_params.py` from their respective exact source checkout. The model was reloaded between legs.

The model tokenizer encodes canonical `<END>` as `[30, 16860, 32]` (3 tokens), but the same text can be emitted as `[30, 39, 48, 38, 32]` (`<`, `E`, `N`, `D`, `>`, 5 tokens). I used SGLang's native `/generate` endpoint with `regex: "<END>"` and negative logit bias for every multi-character token that could encode a substring of the marker. This deterministically made the real model take the 5-token path while leaving stop handling to the server.

<details>
<summary>Identical request used for both legs</summary>

```json
{
  "text": "Output the exact marker <END> and nothing else.",
  "sampling_params": {
    "temperature": 0,
    "max_new_tokens": 8,
    "stop": "<END>",
    "regex": "<END>",
    "logit_bias": {
      "2389": -100.0,
      "4928": -100.0,
      "16860": -100.0,
      "85911": -100.0
    }
  },
  "return_logprob": true,
  "return_text_in_logprobs": true
}
```

</details>

Before (baseline):

```json
{
  "text": "<END>",
  "output_ids": [30, 39, 48, 38, 32, 1],
  "finish_reason": {"type": "stop", "matched": 1}
}
```

The baseline missed the string stop, leaked `<END>`, and continued to EOS (`1`).

After (patch):

```json
{
  "text": "",
  "output_ids": [30, 39, 48, 38, 32],
  "finish_reason": {"type": "stop", "matched": "<END>"}
}
```

The patch saw the same five generated marker tokens, matched the string stop, and trimmed it. A strict offline judge confirmed: identical requests, identical five-token marker path, baseline leak, patched trim, and patched string-stop match all passed.

Limitation: I also replayed the issue's exact English and Chinese chat prompts on both revisions. In this single-replica B300 setup, both revisions naturally used a canonical stop-token path (`banana` / `香蕉`, `matched_stop: "<END>"`), so that request did not reproduce the reported H20 PD-disaggregated symptom. The controlled A/B above validates the decode-window invariant with the actual model and exact source revisions, but is not a reproduction of the original PD topology.

**Comment by DylanB03** [comment]
@hnyls2002 Could you please review this stop-string sampling fix and enable the appropriate CI when convenient? It fixes #36698. The focused sampling unit file passes locally (70 tests and 13 subtests), and the exact DeepSeek-V4-Flash-0731 base/patch A/B evidence is [above](https://github.com/sgl-project/sglang/pull/36728#issuecomment-5446871446). The current red checks are the expected label gates from the missing `run-ci` label, not executed test failures.

**Comment by shaoyuyoung** [comment]
hi, any update here? no reviewers come to review this PR?
